ffi/dotnet/Devolutions.IronRdp/src/WebsocketStream.cs

```diff
@@ -0,0 +1,64 @@
+﻿using System;
+using System.Diagnostics;
+using System.IO;
+using System.Net.WebSockets;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Devolutions.IronRdp.src
+{
+    public class WebsocketStream : Stream
+    {
+        private readonly ClientWebSocket _webSocket;
+
+        public WebsocketStream(ClientWebSocket webSocket)
+        {
+            _webSocket = webSocket;
+        }
+
+        public override bool CanRead => true;
+        public override bool CanSeek => false;
+        public override bool CanWrite => true;
+        public override long Length => throw new NotSupportedException();
+        public override long Position { get => throw new NotSupportedException(); set => throw new NotSupportedException(); }
+
+        public override async Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            var result = await _webSocket.ReceiveAsync(
+                new ArraySegment<byte>(buffer, offset, count), cancellationToken);
+
+            if (result.CloseStatus.HasValue)   // remote sent a CLOSE frame
+            {
+                return 0;                      // treat as end-of-stream
+            }
+
+            return result.Count;
+        }
+
+        public override async Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+        {
+            if (count == 0)
+            {
+                // Note: this is particularly important, if we send a zero-length frame,
+                // somehow Gateway will raise TLS issue during the proxy.
+                return; // Nothing to write
```

Comment:
note: I’m not sure why you get a TLS issue at this point, since you are supposed to have an already established TLS connection at this point, but `SendAsync` with an empty ArraySegment will result in a WebSocket frame to be send. A "Binary" frame that contains an empty byte array.